palace/models/timeoperator.cpp

```diff
@@ -120,10 +108,12 @@ class TimeDependentFirstOrderOperator : public mfem::TimeDependentOperator
     u2.UseDevice(true);
     rhs1.UseDevice(true);
     rhs2.UseDevice(true);
-    u.GetSubVector(idx1, u1);
-    u.GetSubVector(idx2, u2);
-    rhs.GetSubVector(idx1, rhs1);
-    rhs.GetSubVector(idx2, rhs2);
+    u.Read();
+    u1.MakeRef(const_cast<Vector &>(u), 0, size_E);
+    u2.MakeRef(const_cast<Vector &>(u), size_E, size_E);
+    rhs.ReadWrite();
+    rhs1.MakeRef(rhs, 0, size_E);
+    rhs2.MakeRef(rhs, size_E, size_E);
```

Comment:
Nice! The `const_cast` isn't my favorite here, as it means the `u1` and `u2` aren't reflecting the safety constraints on `u`. I have a trick I think should work to address this using structured binding:
```diff
diff --git a/palace/models/timeoperator.cpp b/palace/models/timeoperator.cpp
index ba5eb184..e12266bc 100644
--- a/palace/models/timeoperator.cpp
+++ b/palace/models/timeoperator.cpp
@@ -18,6 +18,28 @@ namespace palace
 namespace
 {
 
+// Helper method for assembling a pair of vectors as references to each half of a vector.
+std::pair<Vector, Vector> AssemblePairedRef(Vector &x, int size)
+{
+  Vector x1, x2;
+  x1.UseDevice(true);
+  x2.UseDevice(true);
+  x1.MakeRef(x, 0, size);
+  x2.MakeRef(x, size, size);
+  return {x1, x2};
+};
+
+// Helper method for assembling a pair of vectors as references to each half of a vector.
+std::pair<const Vector, const Vector> AssemblePairedRef(const Vector &x, int size)
+{
+  Vector x1, x2;
+  x1.UseDevice(true);
+  x2.UseDevice(true);
+  x1.MakeRef(const_cast<Vector &>(x), 0, size);
+  x2.MakeRef(const_cast<Vector &>(x), size, size);
+  return {x1, x2};
+};
+
 class TimeDependentFirstOrderOperator : public mfem::TimeDependentOperator
 {
 public:
@@ -103,17 +125,8 @@ public:
   // Form the RHS for the first-order ODE system
   void FormRHS(const Vector &u, Vector &rhs) const
   {
-    Vector u1, u2, rhs1, rhs2;
-    u1.UseDevice(true);
-    u2.UseDevice(true);
-    rhs1.UseDevice(true);
-    rhs2.UseDevice(true);
-    u.Read();
-    u1.MakeRef(const_cast<Vector &>(u), 0, size_E);
-    u2.MakeRef(const_cast<Vector &>(u), size_E, size_E);
-    rhs.ReadWrite();
-    rhs1.MakeRef(rhs, 0, size_E);
-    rhs2.MakeRef(rhs, size_E, size_E);
+    const auto [u1, u2] = AssemblePairedRef(u, size_E);
+    auto [rhs1, rhs2] = AssemblePairedRef(rhs, size_E);
 
     // u1 = Edot, u2 = E
     // rhs1 = -(K * u2 + C * u1) - J(t)
@@ -140,17 +153,8 @@ public:
     }
     FormRHS(u, RHS);
 
-    Vector du1, du2, RHS1, RHS2;
-    du1.UseDevice(true);
-    du2.UseDevice(true);
-    RHS1.UseDevice(true);
-    RHS2.UseDevice(true);
-    du.ReadWrite();
-    du1.MakeRef(du, 0, size_E);
-    du2.MakeRef(du, size_E, size_E);
-    RHS.ReadWrite();
-    RHS1.MakeRef(RHS, 0, size_E);
-    RHS2.MakeRef(RHS, size_E, size_E);
+    auto [du1, du2] = AssemblePairedRef(du, size_E);
+    auto [RHS1, RHS2] = AssemblePairedRef(RHS, size_E);
 
     kspM->Mult(RHS1, du1);
     du2 = RHS2;
@@ -171,17 +175,8 @@ public:
     Mpi::Print("\n");
     FormRHS(u, RHS);
 
-    Vector k1, k2, RHS1, RHS2;
-    k1.UseDevice(true);
-    k2.UseDevice(true);
-    RHS1.UseDevice(true);
-    RHS2.UseDevice(true);
-    k.ReadWrite();
-    k1.MakeRef(k, 0, size_E);
-    k2.MakeRef(k, size_E, size_E);
-    RHS.ReadWrite();
-    RHS1.MakeRef(RHS, 0, size_E);
-    RHS2.MakeRef(RHS, size_E, size_E);
+    auto [k1, k2] = AssemblePairedRef(k, size_E);
+    auto [RHS1, RHS2] = AssemblePairedRef(RHS, size_E);
 
     // A k1 = RHS1 - dt K RHS2
     K->AddMult(RHS2, RHS1, -dt);
@@ -215,18 +210,11 @@ public:
   // Solve (Mass - dt Jacobian) x = Mass b
   int SUNImplicitSolve(const Vector &b, Vector &x, double tol) override
   {
-    Vector b1, b2, x1, x2, RHS1;
-    b1.UseDevice(true);
-    b2.UseDevice(true);
-    x1.UseDevice(true);
-    x2.UseDevice(true);
+    const auto [b1, b2] = AssemblePairedRef(b, size_E);
+    auto [x1, x2] = AssemblePairedRef(x, size_E);
+
+    Vector RHS1;
     RHS1.UseDevice(true);
-    b.Read();
-    b1.MakeRef(const_cast<Vector &>(b), 0, size_E);
-    b2.MakeRef(const_cast<Vector &>(b), size_E, size_E);
-    x.ReadWrite();
-    x1.MakeRef(x, 0, size_E);
-    x2.MakeRef(x, size_E, size_E);
     RHS.ReadWrite();
     RHS1.MakeRef(RHS, 0, size_E);
```

If that works, we might actually want to use the same trick in other places, then the `const` violation will be trapped to one relatively safe location.